### PR TITLE
Speaker embedding extractor (with ESPnet pre-trained speaker model)

### PR DESCRIPTION
**Pull request by ftshijt**
## What?

- Add ESPnet speaker embedding extractor (inference script)
- Add ESPnet speaker embedding extractor for TTS purpose
- Separate the spk embedding and id converting stage in TTS
    - For flexibility concerns (e.g., after formatting the waveform to use different speaker embedding)
- change xvector to spk_embed as suggested by @Jungjee 

## TODO
- [ ] replace the current spk.sh template with the new inference
- [x] upload pre-trained vctk model with the ESPnet speaker pre-trained model
- [x] test function for spk_inference.py

### CODE REVIEW

**Comment by ftshijt** [comment]
@Jungjee Please feel free to have a check for the implementation~

File: egs2/TEMPLATE/tts1/tts.sh [hunk 1]
```diff
@@ -322,6 +326,7 @@ if ! "${skip_data_prep}"; then
 
         log "Stage 2: Format wav.scp: data/ -> ${data_feats}/"
         for dset in "${train_set}" "${valid_set}" ${test_sets}; do
+            continue
```
Comment (Fhrozen) [inline egs2/TEMPLATE/tts1/tts.sh:329]: Do not forget to remove this after final submit.

Comment (ftshijt) [reply egs2/TEMPLATE/tts1/tts.sh]: Oh, many thanks for noting it!

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py [hunk 2]
```diff
@@ -81,8 +81,37 @@ def __init__(self, args, device):
...
+                )
+                model_tag = args.pretrained_model
+
+            self.speech2embedding = Speech2Embedding.from_pretrained(
+                model_tag=model,
+                **speech2embedding_kwargs,
+            )
 
-    def rawnet_extract_embd(self, audio, n_samples=48000, n_segments=10):
+    def _rawnet_extract_embd(self, audio, n_samples=48000, n_segments=10):
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:114]: maybe we can remove this now? 
Okay to move it later through a different PR. But our RawNet now outperforms this one (which I uploaded in my personal repo previously)

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py [hunk 3]
```diff
@@ -102,13 +131,26 @@ def rawnet_extract_embd(self, audio, n_samples=48000, n_segments=10):
...
         return output.mean(0).detach().cpu().numpy()
 
+    def _espnet_extract_embd(self, audio):
+        if len(audio.shape) > 1:
+            raise ValueError(
+                "Not support multi-channel input for ESPnet pre-trained model"
+                f"Input data has a shape of {audio.shape}."
+            )
+        audio = torch.from_numpy(audio.astype(np.float32)).to(self.device)
+        output = self.self.speech2embedding(audio)
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:141]: Like L150-152, how about giving a warning and averaging it to a single channel? 
In terms of spk recognition performance, this is normally ok. 

But I'm not sure about SE and TTS.

File: egs2/TEMPLATE/spk1/setup.sh [hunk 4]
```diff
@@ -41,7 +41,7 @@ done
 
 
 # Symlinks to TEMPLATE/spk1
-for f in sv.sh path.sh scripts pyscripts; do
+for f in spk.sh path.sh scripts pyscripts; do
```
Comment (Jungjee) [inline egs2/TEMPLATE/spk1/setup.sh:44]: Thx!

File: egs2/TEMPLATE/tts1/tts.sh [hunk 5]
```diff
@@ -67,8 +67,10 @@ f0max=400 # Minimum f0 for pitch extraction.
 
 # X-Vector related
 use_xvector=false   # Whether to use x-vector.
-xvector_tool=kaldi  # Toolkit for extracting x-vector (speechbrain, rawnet, espnet, kaldi)
-xvector_model=speechbrain/spkrec-ecapa-voxceleb  # For only espnet, speechbrain, or rawnet
+xvector_tag=        # The tag of xvector folder.
+xvector_gpu_inference=false # Whether to use gpu to inference xvector.
+xvector_tool=kaldi  # Toolkit for extracting x-vector (speechbrain, rawnet, espnet, kaldi).
+xvector_model=speechbrain/spkrec-ecapa-voxceleb  # For only espnet, speechbrain, or rawnet.
```
Comment (Jungjee) [inline egs2/TEMPLATE/tts1/tts.sh:73]: consider the first comment

**Review by Jungjee — COMMENTED** [review]
Thanks for your effort @ftshijt !! I left some comments. 
Mostly looks good to me.

**Comment by Jungjee** [comment]
One more small comment/question. 
Is there a good place to guide people how to extract speaker embeddings using the new script of yours? 

Maybe in `egs2/TEMPLATE/spk1/README.md`? I'm not sure what is the convention on this.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py [hunk 6]
```diff
@@ -81,8 +81,37 @@ def __init__(self, args, device):
                 )["model"]
             )
             self.model.to(device).eval()
+        elif self.toolkit == "espnet":
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:84]: Sounds good! The only reason I keep the xvector name is to keep it compatible with previous TTS checkpoints. I can use the new name, but would need a bit more time to check the compatibility.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py [hunk 7]
```diff
@@ -81,8 +81,37 @@ def __init__(self, args, device):
...
+            from espnet2.bin.spk_inference import Speech2Embedding
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:85]: Okay, no problem for this.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py (hunk 2 above)
Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:114]: I would suggest keeping it as a legacy api for previous checkpoints

Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:114]: We can always suggest to use our new model.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py (hunk 3 above)
Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:141]: Sounds good, it should be fine.

**Comment by ftshijt** [comment]
> One more small comment/question. Is there a good place to guide people how to extract speaker embeddings using the new script of yours?
> 
> Maybe in `egs2/TEMPLATE/spk1/README.md`? I'm not sure what is the convention on this.

For using it in TTS, we have `egs2/TEMPLATE/tts1/README.md`. As for spk, I guess we can also prepare an spk readme in a similar style. A complementary way would be to use a colab as the demonstration.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sasv-challenge) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:84]: okay! thanks!

[the remark this answers is not in the document]
Comment (Jungjee) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_xvectors.py:84]: okay! thx

**Comment by ftshijt** [comment]
@Jungjee Hi, I applied the name change and verified most of the process (but still need double-check for some previous checkpoints of TTS). But I believe it would be ready for review.

Btw, for an example usage of the API:


```
from espnet2.bin.spk_inference import Speech2SpkEmbedding
import numpy as np

# for huggingface
speech2spk_embed = Speech2SpkEmbedding.from_pretrained(model_tag="espnet/voxcelebs12_rawnet3")
speech2spk_embed(np.zeros( 16500))

# for local ckpt
speech2spk_embed = Speech2SpkEmbedding(model_file="model.pth", train_config="config.yaml")
speech2spk_embed(np.zeros(32000))
```

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_spk_embed.py [hunk 8]
```diff
@@ -102,13 +137,29 @@ def rawnet_extract_embd(self, audio, n_samples=48000, n_segments=10):
...
     def __call__(self, wav, in_sr):
         if self.toolkit == "speechbrain":
             wav = self.audio_norm(torch.from_numpy(wav), in_sr).to(self.device)
             embeds = self.model.encode_batch(wav).detach().cpu().numpy()[0]
         elif self.toolkit == "rawnet":
             wav = librosa.resample(wav, orig_sr=in_sr, target_sr=16000)
-            embeds = self.rawnet_extract_embd(wav)
+            embeds = self._rawnet_extract_embd(wav)
+        elif self.toolkit == "espnet":
+            wav = librosa.resample(wav, orig_sr=in_sr, target_sr=16000)
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/extract_spk_embed.py:161]: how about this?
```
if in_sr != 16000 (or trg_sr):
    wav = librosa.resample(wav, orig_sr=in_sr, target_sr=16000(or trg_sr)
```

**Review by Jungjee — APPROVED** [review]
LGTM! 
Just have one minor suggestion.

File: egs2/TEMPLATE/asr1/pyscripts/utils/extract_spk_embed.py (hunk 8 above)
Comment (ftshijt) [reply egs2/TEMPLATE/asr1/pyscripts/utils/extract_spk_embed.py:161]: Sounds good

**Comment by Jungjee** [comment]
> One naming-level comment.
> How about changing the class name from `Speech2SpkEmbedding` to `Speech2Embedding`?
> We may also provide possible other embedding vectors (e.g., lang or whatever) with the same API name.
> `Speech2Text` is based on this policy (it would be ASR or OWSM S2T).

I see, I didn't think about that.
I think it's a good suggestion! @ftshijt, sorry let's go with your first choice !

**Comment by sw005320** [comment]
LGTM.
Is it ready for merge?

`replace the current spk.sh template with the new inference` in TODO is not checked yet.

**Comment by ftshijt** [comment]
> LGTM. Is it ready for merge?
> 
> `replace the current spk.sh template with the new inference` in TODO is not checked yet.

Sorry, it is not done yet. I recently mostly focused on checking the TTS performance (which is good). Will back to that later this week.

**Comment by sw005320** [comment]
Sounds good.
Please ping me if you finish it.

**Comment by Jungjee** [comment]
> # for huggingface
> speech2spk_embed = Speech2SpkEmbedding.from_pretrained(model_tag="espnet/voxcelebs12_rawnet3")
> speech2spk_embed(np.zeros( 16500))

FYI, to me, replacing existing stage 6 with this extraction can be done in another PR since this can impact the speed of current inference on models and also need several tests. 
(I'm bit worried about losing easy multi-GPU extraction that I've currently made. To not lose speed and at the same time use the new HF-based extraction, to me quite a lot of codes need to be fixed)

**Comment by sw005320** [comment]
@ftshijt
Given the discussion with @Jungjee, it would be good to split the PR about "replace the current spk.sh template with the new inference"
So, I just merged this PR.
Thanks for your great PR and please continue it with the other PR!

**Pull request merged**